keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,450 @@
+
+### Version Skew Strategy
+
+- As for the difference between in-tree components and out-of-tree components, they work independently, so they won't affect each other,
+- As for in-tree components, the protection logic involves only in-tree admission controller and in-tree secret-protection-controller, so version skew won't happen unless these components are available with different versions,
+- As for out-of-tree components, the protection logic involves only out-of-tree admission controller and out-of-tree secret-protection-volumesnapshot-controller, so version skew won't happen unless these components are available with different versions,
+- As for resources, CSI Volume and CSI Snapshot are involved, changes in the API/CRD of these resources especially for their secret fields might cause issue. Howerver, this should be compatibility issue for these API/CRDs.
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: SecretInUseProtection
+  - Components depending on the feature gate:
+    - kube-controller-manager (in-tree)
+    - secret-protection-controllerr (in-tree)
+    - storageobjectinuseprotection admission plugin (in-tree)
+    - secret-protection-volumesnapshot-controller (out-of-tree)
+    - storageobjectinuseprotection volumesnapshot admission plugin (out-of-tree)
+
+###### Does enabling the feature change any default behavior?
+
+Secrets aren't deleted until the resources using them are deleted.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, by disabling the feature gate. After the feature gate is disabled, users need to manually delete the `kubernetes.io/secret-protection` finalizer and the `snapshot.storage.kubernetes.io/secret-protection` finalizer to make secret deleted properly.
```

Comment:
I would not recommend requiring users to manually clean up finalizers.

Reply:
Any ideas to work around? Adding command/script for the purpose?
(Actually, pv-protection/pvc-protection have been there without it).

Reply:
Added my ideas in `Design Details` section. Any feedback is welcome. Thank you for raising this.

Reply:
@liggitt 

I considered a bit deeper on in which case the users are forced to delete finalizer manually. With annotation workaround, users won't need to do it if the controller exist. As for in-tree, the controller has a feature to force delete finalizer if feature is disabled, and the controller exists at least one version as alpha. Therefore this happens only below two cases:
1. Users enabled this feature via feature gate while alpha, and downgrade 1+ version where this controller doesn't exist,  
2. After this feature become beta, and downgrade 2+ version,

It seems that it's very rare to hit this condition.

As for out-of-tree, the same thing could be said, if controller has a nob to disable this feature. Therefore, I'm thinking about adding some kind of feature flag to enable/disable this feature to out-of-tree controller (It doesn't necessary be the same flag with in-tree).

WDYT?

Reply:
Just to be clear, I've made change "adding some kind of feature flag to enable/disable this feature to out-of-tree controller" described in the above comment as `--ephemeral-protection=enabled` flag. If it won't be necessary, I will remove it. Please give feedback on it.